### ISSUE TITLE
scss-lint 0.41 linters fail with undefined method when parsing some string interpolations

### ISSUE THREAD

**Issue by JohnAlbin** [brigade/scss-lint#568]
I just upgraded to scss-lint 0.41 and the entire scss-lint command fails with "undefined method" when parsing my scss files.

It seems to be related to string interpolation. I've narrowed it down to these test cases.

**Test 1**
```
/*
#{hue(#639)}
*/
```

Results:
```
SCSSLint::Linter::ColorKeyword raised unexpected error linting file test1.scss: 'undefined method `[]' for nil:NilClass'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter/color_keyword.rb:8:in `visit_script_color'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:36:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:25:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:49:in `run_linter'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:30:in `block in find_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:28:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:28:in `find_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:18:in `block in run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:17:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:17:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:58:in `scan_for_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:51:in `act_on_options'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:28:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/bin/scss-lint:6:in `<top (required)>'
ruby/2.2.0/bin/scss-lint:23:in `load'
ruby/2.2.0/bin/scss-lint:23:in `<main>'
Report this bug at https://github.com/brigade/scss-lint/issues

To help fix this issue, please include:
- The above stack trace
- SCSS-Lint version: 0.41.0
- Sass version: 3.4.18
- Ruby version: 2.2.3
```

**Test 2**
```
// scss-lint:disable comment
/* #{'string'} */
```

Results:
```
SCSSLint::Linter::SpaceBetweenParens raised unexpected error linting file test2.scss: 'undefined method `[]' for nil:NilClass'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter/space_between_parens.rb:80:in `feel_for_enclosing_parens'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter/space_between_parens.rb:22:in `feel_for_parens_and_check_node'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:36:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:25:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:49:in `run_linter'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:30:in `block in find_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:28:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:28:in `find_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:18:in `block in run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:17:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:17:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:58:in `scan_for_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:51:in `act_on_options'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:28:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/bin/scss-lint:6:in `<top (required)>'
ruby/2.2.0/bin/scss-lint:23:in `load'
ruby/2.2.0/bin/scss-lint:23:in `<main>'
Report this bug at https://github.com/brigade/scss-lint/issues

To help fix this issue, please include:
- The above stack trace
- SCSS-Lint version: 0.41.0
- Sass version: 3.4.18
- Ruby version: 2.2.3
```

**Test 3**
```
$color: rebeccapurple;

/* #{hue($color)} */
```

Results:
```
SCSSLint::Linter::SpaceBetweenParens raised unexpected error linting file test3.scss: 'undefined method `count' for nil:NilClass'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter/space_between_parens.rb:105:in `trim_right_paren'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter/space_between_parens.rb:38:in `check'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter/space_between_parens.rb:9:in `check_node'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:36:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:142:in `block in visit_children'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:140:in `visit_children'
ruby/2.2.0/gems/sass-3.4.18/lib/sass/tree/visitors/base.rb:38:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:131:in `visit'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/linter.rb:25:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:49:in `run_linter'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:30:in `block in find_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:28:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:28:in `find_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:18:in `block in run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:17:in `each'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/runner.rb:17:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:58:in `scan_for_lints'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:51:in `act_on_options'
ruby/2.2.0/gems/scss_lint-0.41.0/lib/scss_lint/cli.rb:28:in `run'
ruby/2.2.0/gems/scss_lint-0.41.0/bin/scss-lint:6:in `<top (required)>'
ruby/2.2.0/bin/scss-lint:23:in `load'
ruby/2.2.0/bin/scss-lint:23:in `<main>'
Report this bug at https://github.com/brigade/scss-lint/issues

To help fix this issue, please include:
- The above stack trace
- SCSS-Lint version: 0.41.0
- Sass version: 3.4.18
- Ruby version: 2.2.3
```

The error messages mention 2 different linters, but I suspect it is some common, underlying parsing code that is failing.

**Comment by srawlins** [comment]
Not good. We're getting a bad`source_range` (3:25 to 3:36) for the Funcall node. Note that that range doesn't exist. I can't tell if an scss-lint monkey patch has broken this range, or Sass (I suspect Sass).

With this patch:

```diff
$ git diff
diff --git i/lib/scss_lint/linter/space_between_parens.rb w/lib/scss_lint/linter/space_between_parens.rb
index 3a47173..db0beeb 100644
--- i/lib/scss_lint/linter/space_between_parens.rb
+++ w/lib/scss_lint/linter/space_between_parens.rb
@@ -6,6 +6,10 @@ module SCSSLint
     include LinterRegistry

     def check_node(node)
+      puts "NODE:     #{node}"
+      puts "INSPECT: `#{node.inspect}`"
+      puts "SOURCE:  `#{source_from_range(node.source_range)}`"
+      puts "AT:      #{node.source_range.inspect}"
       check(node, source_from_range(node.source_range))
       yield
     end
```

on your last test:

```
$ cat issue-568-2.scss
$color: rebeccapurple;

/* #{hue($color)} */
```

I see the following:

```
$ bundle exec bin/scss-lint issue-568-2.scss
NODE:     #<Sass::Script::Tree::Funcall:0x007fbc42e85070>
INSPECT: `hue($color)`
SOURCE:  ``
AT:      (3:25 to 3:36 in issue-568-2.scss)
SCSSLint::Linter::SpaceBetweenParens raised unexpected error linting file issue-568-2.scss: 'undefined method `count' for nil:NilClass'
/Users/srawlins/code/scss-lint/lib/scss_lint/linter/space_between_parens.rb:109:in `trim_right_paren'
/Users/srawlins/code/scss-lint/lib/scss_lint/linter/space_between_parens.rb:42:in `check'
/Users/srawlins/code/scss-lint/lib/scss_lint/linter/space_between_parens.rb:13:in `check_node'
...
```

I'm using sass 3.4.17 (reporter is using 3.4.18)

**Comment by storypixel** [comment]
Getting something similar fwiw

**Comment by sds** [comment]
Duplicate of #531. 

I've implemented a workaround that avoids linting interpolation in `/*...*/` comments since the Sass parser returns incorrect source ranges, and I think for most developers it's not critical that this code be linted.

It looks like @srawlins has a pull request in https://github.com/sass/sass/pull/1875 which fixes the upstream issue, but given that might take a while to merge and eventually released in a gem version I'm inclined to go with the workaround I've implemented in 27492d33322f87c3d47420db39cb39cddcc67983 until then.
